src/Microsoft.Health.Fhir.Liquid.Converter/Filters/SegmentFilters.cs

```diff
@@ -128,5 +128,47 @@ public static bool HasSegments(Hl7v2Data hl7v2Data, string segmentIdContent)
 
             return result;
         }
+
+        public static List<Hl7v2Data> SliceDataBySegments(Hl7v2Data hl7v2Data, List<Hl7v2Segment> segmentList, Hl7v2Segment endSegment = null)
+        {
+            var results = new List<Hl7v2Data>();
+            var result = new Hl7v2Data();
+            var startSegmentFound = false;
+            var segmentIndex = 0;
+            for (var i = 0; i < hl7v2Data.Meta.Count; ++i)
```

Comment:
i feel the implementation logic is not very clear, maybe it's not the best case to using 'for' here?

Reply:
I think if we use "for" here, we can traverse the data only once to reduce the complexity. Although it may seem obscure, the logic is relatively simple, that is, judge whether to find the starting segment and when to terminate the traversal.